### DISCUSSION TITLE
How does HOROVOD_FUSION_THRESHOLD work?

### DISCUSSION THREAD

**Discussion by ihchoi12** [horovod/horovod#3554] [Q&A]
According to the Horovod [documentation](https://horovod.readthedocs.io/en/stable/tensor-fusion_include.html), the default fusion buffer size is 128 MB. I'm running ResNet50 model training on two AWS p3.2xlarge nodes, and [ResNet50's model size](https://keras.io/api/applications/) is 98 MB. Then, does it mean that the entire parameters should be batched into a single AllReduce operation? However, in my experiment, I see a lot of HorovodAllreduce operations are invoked in each iteration as shown in the Nsight timeline below:
<img width="1004" alt="image" src="https://user-images.githubusercontent.com/37887404/170943517-a8d77ff5-1e19-4a52-bc7d-f8b9d8e20ab9.png">

Could you point me out if I have some misunderstanding?

**Comment by maxhgerlach** [comment]
Hi @ihchoi12, all those `HorovodAllreduce` ops that you see there in NVTX are basically framework level. For each gradient allreduce your code requests, one of these ops is called. But the Horovod background thread will fuse the payloads of these individual operations together so that they will be transferred like one fused tensor. I think, if you open up the detailed view for your GPU, you should be able to see that there is one actual NCCL allreduce incorporating multiple tensors.

You could also enable Horovod timeline, then you will get more detailed annotations in NSight Systems, too.

**Reply by ihchoi12** [reply]
Hi @maxhgerlach, thanks for your reply!
I have some follow-up questions here.
1. I still see quite many number of NCCL allreduce operations invoked in the detailed GPU stream view (screenshot below). Doesn't it mean that there are multiple allreduce operations are perforemd by NCCL?
2. Can we enable the horovod timeline inside the Nsight Systems, not the Chrome tracing view?
3. What does NCCL allreduce operation actually do here? Does it send the DMA write request to DMA controller to write the gradient in GPU memory to host memory? 

<img width="1036" alt="image" src="https://user-images.githubusercontent.com/37887404/170987980-50e08b8d-7809-4813-a130-c9aa9962776d.png">

It's still quite confusing to me.. Hope you could kindly help for me to clarify it. Thanks!

**Reply by maxhgerlach** [reply]
> I still see quite many number of NCCL allreduce operations invoked in the detailed GPU stream view (screenshot below). Doesn't it mean that there are multiple allreduce operations are perforemd by NCCL?

Yes, but not one per tensor, so fusion is working. Are some of those for different data types maybe? Or do you have some non-allreduce operations in there as well?

You could also tune knobs like `HOROVOD_FUSION_THRESHOLD` and `HOROVOD_CYCLE_TIME` to influence the tensor fusion behavior if you feel the need. You could also try the Autotune mechanism, but I personally haven't ever.

> Can we enable the horovod timeline inside the Nsight Systems, not the Chrome tracing view?

Yes, you will also get equivalent data in Nsight Systems if you enable writing the tracings that you can view in Chrome. See this PR description for some notes #2723 (has been merged and released for a while now, but I don't think there's better documentation yet...).

> What does NCCL allreduce operation actually do here? Does it send the DMA write request to DMA controller to write the gradient in GPU memory to host memory?

That's very dependent on your GPU, mainboard, and networking. If you sign up with Nvidia, you could get access to this recent talk which describes what could be going on under the hood:
https://www.nvidia.com/en-us/on-demand/session/gtcspring22-s41784/

**Reply by ihchoi12** [reply]
Thanks @maxhgerlach, the Nvidia talk is very informative. 

Speaking of the enabling horovod timeline into the Nsight systems, since the PR is already merged and released, I should be able to see it in the Nsight Systems when I use the latest horovod, isn't it? I generated the above Nsight Systems profile using the latest horovod docker container, but I don't see the timeline in it. Is there anything I need to do to see it? 

FYI, this is the command I use:
`
horovodrun -np 2 -H localhost:1,13.250.102.43:1 -p 12345 --verbose nsys profile -t cuda,osrt,nvtx,cudnn,cublas -f true -o ./baseline -w true  python /root/byteps/bytescheduler/examples/pytorch_horovod_benchmark.py --num-iters 2 --model resnet50 --num-classes 10000
`
Thanks!

**Reply by maxhgerlach** [reply]
Hi @ihchoi12, to get the detailed timeline-like annotations you currently need to enable writing the timeline data to a tracing file, too. For example, set an environment variable like `HOROVOD_TIMELINE=/path/to/timeline.json`. Gathering that data adds some runtime overhead, that's why it's optional. Granted, it's a bit confusing that it's coupled to writing that file -- even if you don't want to look at it, but only want to use the profiler. There's always room for improvement. :)
